### PR TITLE
Migrate `include_bytes_deps`, `optimization-remarks-dir-pgo`, `optimization-remarks-dir`, `issue-40535` and `rmeta-preferred` `run-make` tests to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Needs BSD tryjob.

try-job: aarch64-apple
try-job: x86_64-msvc
try-job: armhf-gnu
try-job: test-various

### CODE REVIEW

**Comment by jieyouxu** [comment]
oops
@bors r-

File: tests/run-make/metadata-only-crate-no-ice/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+fn main() {
+    rustc().input("baz.rs").emit("metadata").run();
+    rustc().input("bar.rs").emit("metadata").extern_("baz", "libbaz.rmeta").run();
+    // There should be no internal compiler error message.
+    rustc()
+        .input("foo.rs")
+        .emit("metadata")
+        .extern_("bar", "libbaz.rmeta")
+        .run()
+        .assert_stderr_not_contains("unexpectedly panicked");
```
Comment (jieyouxu) [inline tests/run-make/metadata-only-crate-no-ice/rmake.rs:18]: Suggestion: this could just `assert_exit_code(1)` (ICE is 101).

**Review by jieyouxu — APPROVED** [review]
Thanks! Just a tiny nit, then r=me

File: tests/run-make/metadata-only-crate-no-ice/rmake.rs (hunk 1 above)
Comment (Oneirical) [reply tests/run-make/metadata-only-crate-no-ice/rmake.rs:18]: I think you mean `0`, as this is supposed to succeed, but yes.

Comment (Oneirical) [reply tests/run-make/metadata-only-crate-no-ice/rmake.rs:18]: Actually, it failing at all is bad. I could really just remove the last line.

**Comment by jieyouxu** [comment]
Actually let's a run a couple of try-jobs before full CI.

**Comment by Oneirical** [comment]
> Actually let's a run a couple of try-jobs before full CI.

Sure, I put OSX and MSVC. Let me know if you'd like to add some more.

**Comment by jieyouxu** [comment]
Yeah, r=me if try jobs come back green, cheers

**Comment by Oneirical** [comment]
I wonder how many developer hours across the whole world were lost because of Windows backslash fun.

**Comment by jieyouxu** [comment]
> I wonder how many developer hours across the whole world were lost because of Windows backslash fun.
> 
> Or maybe the filename is different on each platform and I shouldn't have hardcoded it?

Consider using `PathBuf::join` which *should* handle platform differences in terms of path separators

**Comment by Oneirical** [comment]
> Consider using `PathBuf::join` which _should_ handle platform differences in terms of path separators. Or maybe that prof data doesn't exist when on windows... we'll have to try.

Shame, I cooked up something mildly cursed (see the last commit) before I saw your comment. If this works, maybe I can simplify it back down to what you said.

@bors try

**Comment by Oneirical** [comment]
@rustbot review

Tell me if you find the new helper function (meant to replace globbing patterns) to be a little much.

File: src/tools/run-make-support/src/lib.rs [hunk 2]
```diff
@@ -247,6 +247,34 @@ pub fn test_while_readonly<P: AsRef<Path>, F: FnOnce() + std::panic::UnwindSafe>
     success.unwrap();
 }
 
+/// Browse the directory `path` non-recursively and return the first file path which starts with `prefix` and has the
+/// file extension `extension`.
+#[track_caller]
+pub fn find_files_by_prefix_and_extension<P: AsRef<Path>>(
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:253]: Suggestion: *maybe* `shallow_find_files_by_prefix_and_extension` to indicate that this is not recursive?

File: src/tools/run-make-support/src/lib.rs [hunk 3]
```diff
@@ -247,6 +247,34 @@ pub fn test_while_readonly<P: AsRef<Path>, F: FnOnce() + std::panic::UnwindSafe>
...
+    extension: &str,
+) -> Vec<PathBuf> {
+    let mut matching_files = Vec::new();
+    for entry in fs_wrapper::read_dir(path) {
+        let entry = entry.expect("failed to read directory entry.");
+        let path = entry.path();
+
+        if path.is_file()
+            && path.file_name().unwrap().to_str().unwrap().starts_with(prefix)
+            && path.extension().unwrap().to_str().unwrap() == extension
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:265]: Suggestion: minor improvement, I think this works

```suggestion
            && path.extension().is_some_and(|ext| ext == extension)
```

File: src/tools/run-make-support/src/lib.rs [hunk 4]
```diff
@@ -247,6 +247,34 @@ pub fn test_while_readonly<P: AsRef<Path>, F: FnOnce() + std::panic::UnwindSafe>
...
+    prefix: &str,
+    extension: &str,
+) -> Vec<PathBuf> {
+    let mut matching_files = Vec::new();
+    for entry in fs_wrapper::read_dir(path) {
+        let entry = entry.expect("failed to read directory entry.");
+        let path = entry.path();
+
+        if path.is_file()
+            && path.file_name().unwrap().to_str().unwrap().starts_with(prefix)
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:264]: Suggestion:

```suggestion
            && path.file_name().is_some_and(|name| name.to_str().unwrap().starts_with(prefix))
```

File: src/tools/run-make-support/src/lib.rs [hunk 5]
```diff
@@ -247,6 +247,34 @@ pub fn test_while_readonly<P: AsRef<Path>, F: FnOnce() + std::panic::UnwindSafe>
...
+            && path.extension().unwrap().to_str().unwrap() == extension
+        {
+            matching_files.push(path);
+        }
+    }
+
+    if matching_files.is_empty() {
+        panic!("no files found with the given prefix and extension");
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:272]: Discussion: I feel like this should not panic given the function name, i.e. return a `vec![]` and let the caller (the test) figure out what to do in case of empty result?

If you do change this, remember to update the corresponding test to panic on empty result.

**Comment by jieyouxu** [comment]
The new helper looks fine, just some nits. @rustbot author

File: tests/run-make/optimization-remarks-dir-pgo/rmake.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    run("foo");
+    llvm_profdata()
+        .merge()
+        .output("merged.profdata")
+        .input(
+            shallow_find_files("profdata", |path| {
+                has_prefix(path, "default") && has_extension(path, "profraw")
+            })
+            .get(0)
+            .unwrap(),
```
Comment (jieyouxu) [inline tests/run-make/optimization-remarks-dir-pgo/rmake.rs:25]: Question/Suggestion: is the reason why we're doing shallow find here that the profdata filename isn't predictable? In any case, can you pull this out to a temp variable, because I feel like it's a bit nested and hard to follow.

```rs
let profdata_file = shallow_find_files("profdata", |path| {
        has_prefix(path, "default") && has_extension(path, "profraw")
    })
    .get(0)
    .unwrap();
```

File: tests/run-make/optimization-remarks-dir-pgo/rmake.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        .opt()
+        .input("foo.rs")
+        .arg("-Cremark=all")
+        .arg("-Zremark-dir=profiles")
+        .run();
+    // Check that PGO hotness is included in the remark files
+    assert!(
+        !shallow_find_files("profiles", |path| {
+            has_prefix(path, "foo") && has_extension(path, "yaml")
+        })
```
Comment (jieyouxu) [inline tests/run-make/optimization-remarks-dir-pgo/rmake.rs:39]: Suggestion:

```rs
shallow_find_files("profiles", |path| {
    has_prefix(path, "foo") && has_extension(path, "yaml")
})
```

is duplicated in the assert and in the following `for` loop. Can you extract out a common temp variable, let's say

```rs
let remark_files = shallow_find_files("profiles", |path| {
    has_prefix(path, "foo") && has_extension(path, "yaml")
});
```

I think that will be easier to read ^^

File: tests/run-make/optimization-remarks-dir/rmake.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    rustc()
+        .opt()
+        .input("foo.rs")
+        .crate_type("lib")
+        .arg("-Cremark=all")
+        .arg("-Zremark-dir=profiles_all")
+        .run();
+
+    for file in shallow_find_files("profiles_all", |path| {
+        has_prefix(path, "foo") && has_extension(path, "yaml")
```
Comment (jieyouxu) [inline tests/run-make/optimization-remarks-dir/rmake.rs:22]: Suggestion: can you either pull out a temp variable for `shallow_find_files` here and below or add a comment on what the files found are? I assume they are remark files?

**Review by jieyouxu — APPROVED** [review]
Looks good in general, I just have some test readability suggestions.

File: tests/run-make/optimization-remarks-dir-pgo/rmake.rs (hunk 6 above)
Comment (Oneirical) [reply tests/run-make/optimization-remarks-dir-pgo/rmake.rs]: Indeed, it's a long sequence of numbers, and I previously used the hardcoded filename in a previous draft of this test, which caused failures on certain architectures.

File: tests/run-make/optimization-remarks-dir/rmake.rs (hunk 8 above)
Comment (Oneirical) [reply tests/run-make/optimization-remarks-dir/rmake.rs]: I did, and I also added the `not_contains` filter for consistency instead of the if statement checking if `codegen` is not in the filename.

@rustbot review

**Review by jieyouxu — APPROVED** [review]
Cool, let's rerun the try jobs once, if they come back green then r=me!
